lib/RestructureCFG/CycleEquivalenceAnalysis.cpp

```diff
@@ -0,0 +1,554 @@
+//
+// Copyright rev.ng Labs Srl. See LICENSE.md for details.
+//
+
+#include <iterator>
+#include <limits>
+#include <map>
+
+#include "llvm/ADT/DepthFirstIterator.h"
+#include "llvm/ADT/PostOrderIterator.h"
+
+#include "revng/ADT/GenericGraph.h"
+#include "revng/Support/Debug.h"
+#include "revng/Support/IRHelpers.h"
+
+#include "revng-c/RestructureCFG/CycleEquivalenceAnalysis.h"
+
+using namespace llvm;
+using namespace llvm::cl;
+
+// Debug logger.
+Logger<> CycleEquivalenceAnalysisLogger("cycle-equivalence");
+
+template<class GraphT, class GT>
+void CycleEquivalenceAnalysis<GraphT, GT>::insertEdge(EdgeDescriptor E,
+                                                      EdgeLabel *Label,
+                                                      bool IsInverted,
+                                                      BracketDescriptor BD) {
+
+  // We do not want to insert `Fake` edges in the equivalence classes. Such
+  // edges can be both the capping backedges and the additional `exit` ->
+  // `entry` bracket
+  revng_assert(Label->kind() != EdgeKind::Invalid);
+  if (Label->kind() == EdgeKind::Fake)
+    return;
+
+  // We marshal the `EdgeDescriptor` object to unwrap the underlying edge, not
+  // wrapped in any `GenericGraph` node, and we invert the direction of the edge
+  // if are traversing it in the inverse direction (wrt the direction of the
+  // original directed edge) during the current exploration
+  using NakedEdgeDescriptor = CycleEquivalenceClass<NodeT>::EdgeDescriptor;
+  NakedEdgeDescriptor NakedEdge;
+  if (not IsInverted) {
+    NakedEdge = NakedEdgeDescriptor(E.first->getBlock(), E.second->getBlock());
+  } else {
+    NakedEdge = NakedEdgeDescriptor(E.second->getBlock(), E.first->getBlock());
+  }
```

Comment:
In reading this comment, take into consideration also the snippet highlighted by the next comment here https://github.com/revng/revng-c/pull/18#discussion_r1634367342.

I think there might be a misunderstanding. `EdgeDescriptor` is not a class per se, but it is just an "elegant" way to refer to a `std::pair` of nodes. In this context, when we want to pass from a `EdgeDescriptor` to a `NakedEdgeDescriptor`, the focus is on the fact that we want to unwrap the internal node, in this case in order to get the original `NodeT` which referred to the nodes on the original graph, and not the node that we use on the `GenericGraph` which wraps it with other additional information. Therefore, I don't get how implementing a method on the `EdgeDescriptor` can help us unwrap from the element of the pair (and therefore dropping the `getBlock` method), and how it can help us get the `NakedEdgeDescriptor` describer by the inverted edge or not.

Reply:
Ok you can't do a method, but it's disturbing to have this `NakedEdgeDescriptor` because it makes it seem like it's somehow different from a regular `EdgeDescriptor`.
Drop the `Naked` prefix, just do:
```c++
using EdgeDescriptor = CycleEquivalenceClass<NodeT>::EdgeDescriptor;
```

Reply:
Ok, I applied your proposed change [here](https://github.com/revng/revng-c/blob/f23593da9f15288aca2fee41975e6e55dbbbcb3b/lib/RestructureCFG/CycleEquivalenceAnalysis.cpp#L46)

There is [another place](https://github.com/revng/revng-c/blob/f23593da9f15288aca2fee41975e6e55dbbbcb3b/include/revng-c/RestructureCFG/CycleEquivalenceResult.h#L32) where we use the `NakedEdgeDescriptor`. However, there we need both the `EdgeDescriptor` and `NakedEdgeDescriptor` in the same namespace, so we can't just remove the `Naked` prefix. I think it is legit such use of it there

Reply:
Ok, but like I said `Naked` is weird. Basically it means that the edge refers to the original nodes. What about `OriginalNodesEdgeDescriptor`?

Reply:
Sure, changed the name [here](https://github.com/revng/revng-c/blob/b693afb4ef873bb340c581382a45b1c14e3d5b32/include/revng-c/RestructureCFG/CycleEquivalenceResult.h#L34)

Reply:
Also changed the name variable from `NakedEdge` to `OriginalEdge` https://github.com/revng/revng-c/blob/b69cf3baad6602473085be7a84750763da0e2b02/lib/RestructureCFG/CycleEquivalenceAnalysis.cpp#L46

Reply:
:+1: